### PR TITLE
Geo

### PR DESCRIPTION
**Pull request by AndreiKingsley**

### CODE REVIEW

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/layerWrap.kt [hunk 1]
```diff
@@ -20,7 +22,13 @@ internal fun Layer.wrap(
     globalSettings: Map<Aes, Setting>,
 ) {
     val dataset = if (datasetIndex == 0) {
-        null
+        val dataset = datasets[datasetIndex]
+        // can't provide geodata as main dataset
+        if (dataset is GeoSpatialData) {
+            dataset
+        } else {
+            null
```
Comment (zaleslaw) [inline kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/layerWrap.kt:30]: If we return null, is it possible to log here something with warning? or throw an exception if we don't handle null higher

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/layerWrap.kt [hunk 2]
```diff
@@ -36,9 +44,19 @@ internal fun Layer.wrap(
     } else {
         settings
     }
-    val df = (datasets[datasetIndex]).dataFrame()
-    featureBuffer.add(LayerWrapper(this, addGroups, dataset?.wrap(), mappings, settings, groupKeys))
+    val map = if (dataset is GeoSpatialData) {
+        dataset.toSpatialDataset()
+    } else null
```
Comment (zaleslaw) [inline kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/layerWrap.kt:49]: How should be these null handled?

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/data/GeoSpatialData.kt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package org.jetbrains.kotlinx.kandy.letsplot.data
+
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.jetbrains.kotlinx.kandy.ir.data.TableData
+import org.jetbrains.letsPlot.spatial.SpatialDataset
+
+public interface GeoSpatialData: TableData {
```
Comment (zaleslaw) [inline kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/data/GeoSpatialData.kt:7]: Please, generate and fill KDocs

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/util/mergePolygons.kt [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package org.jetbrains.kotlinx.kandy.letsplot.geo.util
+
+import org.jetbrains.kotlinx.dataframe.DataColumn
+import org.jetbrains.kotlinx.dataframe.api.map
+import org.locationtech.jts.geom.Geometry
+import org.locationtech.jts.geom.GeometryFactory
+import org.locationtech.jts.geom.MultiPolygon
+import org.locationtech.jts.operation.union.CascadedPolygonUnion
+
+
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/util/mergePolygons.kt:10]: Remove empty line and add some KDocs, great that exception is raised? Could we throw IllegalArgumentException in other situation which I commented?

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoPolygon.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import kotlin.reflect.typeOf
+
+// TODO add ColumnAccessor & String api
+
+public inline fun GeoDataScope.geoPolygon(
+    block: PolygonBuilder.() -> Unit = {}
+) {
+    geometry().forEach {
+        if (it !is Polygonal) {
+            error("Not a polygon geometry: $it")
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoPolygon.kt:27]: why it somewhere IllegalArgumentException, somewhere error? could it be unified through the PR?

Comment (zaleslaw) [reply kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoPolygon.kt:27]: Also KDocs required

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoPoints.kt [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import org.jetbrains.kotlinx.kandy.letsplot.geo.dsl.geometry
+import org.jetbrains.kotlinx.kandy.letsplot.layers.builders.PointsBuilder
+import org.jetbrains.kotlinx.kandy.letsplot.layers.points
+import org.locationtech.jts.geom.Geometry
+import org.locationtech.jts.geom.MultiPoint
+import org.locationtech.jts.geom.Point
+import org.locationtech.jts.geom.Puntal
+import kotlin.reflect.typeOf
+
+// TODO add ColumnAccessor & String api
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoPoints.kt:20]: TODO: add KDocs

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoDataScope.kt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+@file:Suppress("INVISIBLE_MEMBER")
+
+package org.jetbrains.kotlinx.kandy.letsplot.geo.dsl
+
+public sealed interface GeoDataScope
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoDataScope.kt:5]: Is it really required a separate file?

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/acсessors.kt [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+@PublishedApi
+internal fun GeoDataScope.geometry(): DataColumn<Geometry> {
+    with((this as LayerCreatorScope).datasetBuilder as GeoDataBuilder) {
+        takeColumn("geometry")
+        return buffer[column<Geometry>("geometry")]
+    }
+}
+
+@PublishedApi
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/acсessors.kt:19]: Sorry, is it @PublishedApi and internal in the same time? Is it correct

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoDataFrameScope.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+package org.jetbrains.kotlinx.kandy.letsplot.geo.dsl
+
+import org.jetbrains.kotlinx.dataframe.ColumnsContainer
+import org.jetbrains.kotlinx.dataframe.geo.GeoDataFrame
+import org.jetbrains.kotlinx.dataframe.geo.WithGeometry
+import org.jetbrains.kotlinx.kandy.dsl.internal.LayerCreatorScope
+import org.jetbrains.kotlinx.kandy.dsl.internal.MultiLayerPlotBuilder
+
+public class GeoDataFrameScope<T : WithGeometry>(
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoDataFrameScope.kt:11]: KDocs please

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoDataFramePlotBuilder.kt [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+package org.jetbrains.kotlinx.kandy.letsplot.geo.dsl
+
+import org.jetbrains.kotlinx.dataframe.ColumnsContainer
+import org.jetbrains.kotlinx.dataframe.geo.GeoDataFrame
+import org.jetbrains.kotlinx.dataframe.geo.WithGeometry
+import org.jetbrains.kotlinx.kandy.dsl.internal.DatasetBuilder
+import org.jetbrains.kotlinx.kandy.dsl.internal.dataframe.MultiLayerPlotBuilderImpl
+
+public class GeoDataFramePlotBuilder<T : WithGeometry> @PublishedApi internal constructor(
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoDataFramePlotBuilder.kt:11]: KDocs please

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+import org.jetbrains.kotlinx.kandy.letsplot.data.GeoSpatialData
+import org.jetbrains.letsPlot.spatial.SpatialDataset
+
+internal class GeoData(val geoDataFrame: GeoDataFrame<*>) : GeoSpatialData {
+    override val dataFrame: DataFrame<WithGeometry> = geoDataFrame.df
+
+    override fun toSpatialDataset(): SpatialDataset {
+        with(geoDataFrame) {
+            // TODO encoding precision
+            val geojson = GeometryJSON(10)
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt:19]: Magic Constant  - should be extracted and explained

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+            return SpatialDataset.withGEOJSON(
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt:20]: is it possible to rename to withGeoJson to be more Kotlinish?

File: kandy-geo/build.gradle.kts [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    implementation(libs.jai.core)
+
+    testImplementation(kotlin("test"))
+}
+
+tasks.withType<KotlinCompile>().configureEach {
+    val friendModules = listOf(project(":kandy-api"), project(":kandy-lets-plot"))
+    val jarTasks = friendModules.map { it.tasks.getByName("jar") as Jar }
+    val jarPaths = jarTasks.map { it.archiveFile.get().asFile.absolutePath }
+    (this as BaseKotlinCompile).friendPaths.from(jarPaths)
```
Comment (zaleslaw) [inline kandy-geo/build.gradle.kts:46]: could please add some note or explanation about why do we need friend modules here?

File: kandy-geo/build.gradle.kts [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+repositories {
+    maven("https://repo.osgeo.org/repository/release")
+    mavenCentral()
+}
+
+// https://stackoverflow.com/questions/26993105/i-get-an-error-downloading-javax-media-jai-core1-1-3-from-maven-central
+// jai core dependency should be excluded from geotools dependencies and added separately
+fun ExternalModuleDependency.excludeJaiCore() = exclude("javax.media", "jai_core")
+
+
```
Comment (zaleslaw) [inline kandy-geo/build.gradle.kts:21]: remove blank lines

**Review by zaleslaw — CHANGES_REQUESTED** [review]
It requires more work on KDocs and some questions should be answered also

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/layerWrap.kt (hunk 1 above)
Comment (AndreiKingsley) [reply kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/layerWrap.kt:30]: There is no error here, it is a note that in the Lets-Plot specification it is impossible to pass geodataset as main (default for layers), so if we have a main geodatset in kandy, it should be passed to each layer separately during processing and that's exactly what's going here.

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/layerWrap.kt (hunk 2 above)
Comment (AndreiKingsley) [reply kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/layerWrap.kt:49]: It shouldn't, it's a feature of the Lets-Plot specification, which is described in the previous conversation

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/acсessors.kt (hunk 8 above)
Comment (AndreiKingsley) [reply kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/acсessors.kt:19]: But `@PublishedApi` is applicable only for internal properties :)

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt (hunk 12 above)
Comment (AndreiKingsley) [reply kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt:20]: It's not our API

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoDataScope.kt (hunk 7 above)
Comment (AndreiKingsley) [reply kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoDataScope.kt:5]: why not, plus I added comments to the interface

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/data/GeoSpatialData.kt [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.jetbrains.kotlinx.kandy.ir.data.TableData
+import org.jetbrains.letsPlot.spatial.SpatialDataset
+
+/**
+ * Represents geospatial data wrapper.
+ *
+ * This interface extends the `TableData` interface, providing specific capabilities for handling geospatial data.
+ *
+ * @property dataFrame The data frame containing the geospatial data.
```
Comment (zaleslaw) [inline kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/data/GeoSpatialData.kt:12]: The `DataFrame` (to refer on class), don't use `data frame`

File: kandy-geo/build.gradle.kts (hunk 13 above)
Comment (zaleslaw) [reply kandy-geo/build.gradle.kts:45]: @AndreiKingsley 
add a comment with motivation "what's happened here? and why we did this?"

in this `tasks.withType` snippet

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoLayer.kt [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+) {
+    if (this is GeoDataScope && geometry() == geometry) {
+        return layerFun(builder)
+    }
+    val columnRowsCount = geometry.size
+    val datasetBuilderRowsCount = datasetBuilder.rowsCount()
+    val geometryColumn = geometry named "geometry"
+    val geoDataFrame = if (columnRowsCount != datasetBuilderRowsCount) {
+        // override dataset if datasetBuilder size doesn't match
+        dataFrameOf(geometryColumn).toGeo(null)
```
Comment (zaleslaw) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoLayer.kt:35]: I don't understand this code `dataFrameOf(geometryColumn).toGeo(null)`

**Review by zaleslaw — CHANGES_REQUESTED** [review]
Some moments a little bit unclear, please clarify them

File: kandy-geo/build.gradle.kts (hunk 13 above)
Comment (AndreiKingsley) [reply kandy-geo/build.gradle.kts:45]: I did

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoLayer.kt (hunk 16 above)
Comment (AndreiKingsley) [reply kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoLayer.kt:35]: Creating a new dataframe with single column and casting to GeoDataFrame (witn null crs)

File: kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/dataframe/concatFixed.kt [hunk 17]
```diff
@@ -13,7 +13,7 @@ internal fun <T, R> GroupBy<T, R>.concatFixed(): DataFrame<R> {
     return mapToFrames {
         val rowsCount = group.rowsCount()
         val keyColumns = keyNames.filter { it !in group.columnNames() }.map { keyName ->
-            DataColumn.create(keyName, List(rowsCount) { key[keyName] }, Infer.Type)
+            DataColumn.Companion.createByType(keyName, List(rowsCount) { key[keyName] }, Infer.Type)
```
Comment (devcrocod) [inline kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/dataframe/concatFixed.kt:16]: Why specify Companion?

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+import org.jetbrains.kotlinx.dataframe.geo.geometry
+import org.jetbrains.kotlinx.kandy.letsplot.data.GeoSpatialData
+import org.jetbrains.letsPlot.spatial.SpatialDataset
+
+internal class GeoData(val geoDataFrame: GeoDataFrame<*>) : GeoSpatialData {
+    override val dataFrame: DataFrame<WithGeometry> = geoDataFrame.df
+
+    companion object {
+        // TODO lets-plot default, encoding precision customization
+        const val DEFAULT_PRECISION = 10
```
Comment (devcrocod) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt:18]: Is there a way to change this?
Why is it in a companion object and a constant?

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoLayer.kt (hunk 16 above)
Comment (devcrocod) [reply kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoLayer.kt:35]: This is not quite in Kotlin style. You can define a copy method with changing parameters

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoPolygon.kt [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+ *
+ *     }
+ * }
+ * ```
+ */
+public inline fun GeoDataScope.geoPolygon(
+    block: PolygonBuilder.() -> Unit = {}
+) {
+    geometry().forEach {
+        if (it !is Polygonal) {
```
Comment (devcrocod) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoPolygon.kt:64]: change to `check`

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoRectangles.kt [hunk 20]
```diff
@@ -0,0 +1,388 @@
...
+        xMin.constant(null)
+        yMin.constant(null)
+        xMax.constant(null)
+        yMax.constant(null)
+        block()
+    }
+}
+
+// TODO add ColumnAccessor & String api
+@JvmName("geoRectanglesGeometry")
```
Comment (devcrocod) [inline kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoRectangles.kt:71]: Why do we need JvmName? Is there a conflict somewhere?

**Review by devcrocod — APPROVED** [review]
I left a few comments
Unfortunately, I don't have the opportunity to look more closely later, but overall I approved this

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt (hunk 18 above)
Comment (AndreiKingsley) [reply kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/dsl/GeoData.kt:18]: it's temporary, will design better way to specify it

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoLayer.kt (hunk 16 above)
Comment (AndreiKingsley) [reply kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoLayer.kt:35]: What do you mean?

File: kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoRectangles.kt (hunk 20 above)
Comment (AndreiKingsley) [reply kandy-geo/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/geo/layers/geoRectangles.kt:71]: Yes:
```
Platform declaration clash: The following declarations have the same JVM signature (geoRectangles(Lorg/jetbrains/kotlinx/kandy/dsl/internal/LayerCreatorScope;Ljava/lang/Iterable;Lkotlin/jvm/functions/Function1;)V):
    fun LayerCreatorScope.geoRectangles(geometry: Iterable<Envelope>, block: (RectanglesBuilder) -> Unit = ...): Unit defined in org.jetbrains.kotlinx.kandy.letsplot.geo.layers
    fun LayerCreatorScope.geoRectangles(geometry: Iterable<Geometry>, block: (RectanglesBuilder) -> Unit = ...): Unit defined in org.jetbrains.kotlinx.kandy.letsplot.geo.layers

**Pull request merged**